From dd3de6f095c1a0e8b7c4fff579541287ab2f4fe6 Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jose.dapena@lge.com>
Date: Mon, 22 Jul 2019 21:17:52 +0200
Subject: Do not assume __has_builtin is available

Compilation with GCC breaks as there is a build check using it.
To avoid the problem, we declare the macro HAS_BUILTIN that maps
to __has_builtin if it is defined, or 0 otherwise.

Bug: chromium:819294
Change-Id: I105d566f08baa495fccd553d7b06eb8ce36621ef
Reviewed-on: https://chromium-review.googlesource.com/c/angle/angle/+/1713604
Commit-Queue: Jamie Madill <jmadill@chromium.org>
Reviewed-by: Jamie Madill <jmadill@chromium.org>

diff --git a/src/common/mathutil.h b/src/common/mathutil.h
index bdf8c2a53..798ddba59 100644
--- a/src/common/mathutil.h
+++ b/src/common/mathutil.h
@@ -1254,14 +1254,20 @@ inline constexpr unsigned int UnsignedCeilDivide(unsigned int value, unsigned in
     return (divided + ((value % divisor == 0) ? 0 : 1));
 }
 
+#if defined(__has_builtin)
+#    define ANGLE_HAS_BUILTIN(x) __has_builtin(x)
+#else
+#    define ANGLE_HAS_BUILTIN(x) 0
+#endif
+
 #if defined(_MSC_VER)
 
 #    define ANGLE_ROTL(x, y) _rotl(x, y)
 #    define ANGLE_ROTL64(x, y) _rotl64(x, y)
 #    define ANGLE_ROTR16(x, y) _rotr16(x, y)
 
-#elif defined(__clang__) && __has_builtin(__builtin_rotateleft32) && \
-    __has_builtin(__builtin_rotateleft64) && __has_builtin(__builtin_rotateright16)
+#elif defined(__clang__) && ANGLE_HAS_BUILTIN(__builtin_rotateleft32) && \
+    ANGLE_HAS_BUILTIN(__builtin_rotateleft64) && ANGLE_HAS_BUILTIN(__builtin_rotateright16)
 
 #    define ANGLE_ROTL(x, y) __builtin_rotateleft32(x, y)
 #    define ANGLE_ROTL64(x, y) __builtin_rotateleft64(x, y)
-- 
2.20.1

